### PR TITLE
[SPARK-19876][SS][WIP] OneTime Trigger Executor

### PR DESCRIPTION
**Pull request by ghost**
## What changes were proposed in this pull request?

An additional trigger and trigger executor that will execute a single trigger only. One can use this OneTime trigger to have more control over the scheduling of triggers. 

In addition, this patch requires an optimization to StreamExecution that logs a commit record at the end of successfully processing a batch. This new commit log will be used to determine the next batch (offsets) to process after a restart, instead of using the offset log itself to determine what batch to process next after restart; using the offset log to determine this would process the previously logged batch, always, thus not permitting a OneTime trigger feature.  

## How was this patch tested?

A number of existing tests have been revised. These tests all assumed that when restarting a stream, the last batch in the offset log is to be re-processed. Given that we now have a commit log that will tell us if that last batch was processed successfully, the results/assumptions of those tests needed to be revised accordingly. 

In addition, a OneTime trigger test was added to StreamingQuerySuite, which tests:
- The semantics of OneTime trigger (i.e., on start, execute a single batch, then stop).
- The case when the commit log was not able to successfully log the completion of a batch before restart, which would mean that we should fall back to what's in the offset log.
- A OneTime trigger execution that results in an exception being thrown.

@marmbrus @tdas @zsxwing 

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetCommitLog.scala [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+  override protected def deserialize(in: InputStream): Option[String] = {
+    // called inside a try-finally where the underlying stream is closed in the caller
+    val lines = IOSource.fromInputStream(in, UTF_8.name()).getLines()
+    if (!lines.hasNext) {
+      throw new IllegalStateException("Incomplete log file")
+    }
+    val version = lines.next()
+    if (version != OffsetCommitLog.VERSION) {
+      throw new IllegalStateException(s"Unknown log version: ${version}")
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetCommitLog.scala]: We should make sure the error here is consistent with the work being done in #17070

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetCommitLog.scala [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    out.write('\n')
+
+    // write metadata or void
+    out.write(metadata.getOrElse(OffsetCommitLog.SERIALIZED_VOID).getBytes(UTF_8))
+  }
+}
+
+object OffsetCommitLog {
+  private val VERSION = "v1"
+  private val SERIALIZED_VOID = "-"
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetCommitLog.scala]: Why not make this an empty json object? `{}`

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/Trigger.scala [hunk 3]
```diff
@@ -38,6 +38,26 @@ sealed trait Trigger
...
+ * A trigger that runs a query once then terminates
+ *
+ * Scala Example:
+ * {{{
+ *   df.write.trigger(OneTime)
+ * }}}
+ *
+ * Java Example:
+ * {{{
+ *   df.write.trigger(OneTime.create())
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/Trigger.scala]: I don't think this works?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetCommitLog.scala [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+package org.apache.spark.sql.execution.streaming
+
+import java.io.{InputStream, OutputStream}
+import java.nio.charset.StandardCharsets._
+
+import scala.io.{Source => IOSource}
+
+import org.apache.spark.sql.SparkSession
+
+class OffsetCommitLog(sparkSession: SparkSession, path: String)
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetCommitLog.scala]: Scala doc please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala [hunk 5]
```diff
@@ -284,6 +291,7 @@ class StreamExecution(
               finishTrigger(dataAvailable)
               if (dataAvailable) {
                 // We'll increase currentBatchId after we complete processing current batch's data
+                commitLog.add(currentBatchId, None)
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: I wonder if we should make this async?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala [hunk 6]
```diff
@@ -377,17 +385,25 @@ class StreamExecution(
...
-        availableOffsets = nextOffsets.toStreamProgress(sources)
-        offsetSeqMetadata = nextOffsets.metadata.getOrElse(OffsetSeqMetadata())
-        logDebug(s"Found possibly unprocessed offsets $availableOffsets " +
-          s"at batch timestamp ${offsetSeqMetadata.batchTimestampMs}")
-
-        offsetLog.get(batchId - 1).foreach {
-          case lastOffsets =>
-            committedOffsets = lastOffsets.toStreamProgress(sources)
-            logDebug(s"Resuming with committed offsets: $committedOffsets")
+        currentBatchId = commitLog.getLatest() match {
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: Mind adding a few more comments here.  This logic is getting very dense.  I think that its doing something like the following:
 - finding the max committed batch
 - checking to see if there is a started but uncommitted batch
 - otherwise constructing a new batch

**Review by marmbrus — CHANGES_REQUESTED** [review]
This is great!  Thanks for working on it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala [hunk 7]
```diff
@@ -206,6 +207,12 @@ class StreamExecution(
    */
   val offsetLog = new OffsetSeqLog(sparkSession, checkpointFile("offsets"))
 
+  /**
+   * A log that records the committed batch ids. This is used to check if a batch was committed
+   * on restart, instead of (possibly) re-running the previous batch.
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: nit: "if a batch was committed on restart" sounds like batches are supposed to get committed only on restart. :)

Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: Also, keep the comment generic such that it does mean that not only the previous batch will be re-run. in future we could be rerun multiple batches.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetCommitLog.scala [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+ * line 2: metadata (optional json string)
+ */
+class OffsetCommitLog(sparkSession: SparkSession, path: String)
+  extends HDFSMetadataLog[Option[String]](sparkSession, path) {
+
+  override protected def deserialize(in: InputStream): Option[String] = {
+    // called inside a try-finally where the underlying stream is closed in the caller
+    val lines = IOSource.fromInputStream(in, UTF_8.name()).getLines()
+    if (!lines.hasNext) {
+      throw new IllegalStateException("Incomplete log file")
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetCommitLog.scala]: can you say "incomplete log file in the offset commit log"

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetCommitLog.scala [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    out.write(OffsetCommitLog.VERSION.toString.getBytes(UTF_8))
+    out.write('\n')
+
+    // write metadata or void
+    out.write(metadata.getOrElse(OffsetCommitLog.SERIALIZED_VOID).getBytes(UTF_8))
+  }
+}
+
+object OffsetCommitLog {
+  private val VERSION = 1
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetCommitLog.scala]: Lets be consistent with other logs in writing "v1" for version and not "1"

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala [hunk 10]
```diff
@@ -373,22 +380,66 @@ class StreamExecution(
    *  - currentBatchId
    *  - committedOffsets
    *  - availableOffsets
+   *  The basic structure of this method is as follows:
+   *
+   *  Identify (from the offset log) the offsets used to run the last batch
+   *  IF a last batch exists THEN
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: "a last batch" is grammatically weird .. isnt it?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala [hunk 11]
```diff
@@ -373,22 +380,66 @@ class StreamExecution(
...
+   *    Set the next batch to that last batch
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: may be "set the next batch to be executed as the last recovered batch"

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala [hunk 12]
```diff
@@ -373,22 +380,66 @@ class StreamExecution(
...
+   *    Check the commit log to see which batch was committed last
+   *    IF the last batch was committed THEN
+   *      Call getBatch using the last batch start and end offsets
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: Add the reason regarding why we do this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala [hunk 13]
```diff
@@ -373,22 +380,66 @@ class StreamExecution(
...
-        logInfo(s"Resuming streaming query, starting with batch $batchId")
-        currentBatchId = batchId
         availableOffsets = nextOffsets.toStreamProgress(sources)
         offsetSeqMetadata = nextOffsets.metadata.getOrElse(OffsetSeqMetadata())
-        logDebug(s"Found possibly unprocessed offsets $availableOffsets " +
-          s"at batch timestamp ${offsetSeqMetadata.batchTimestampMs}")
-
-        offsetLog.get(batchId - 1).foreach {
-          case lastOffsets =>
+        if (batchId > 0) {
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: why are we introducing this condition?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala [hunk 14]
```diff
@@ -373,22 +380,66 @@ class StreamExecution(
...
+   *      Call getBatch using the last batch start and end offsets
+   *      Setup for a new batch i.e., start = last batch end, and identify new end
+   *    DONE
+   *  ELSE
+   *    Identify a brand new batch
+   *  DONE
    */
   private def populateStartOffsets(): Unit = {
     offsetLog.getLatest() match {
       case Some((batchId, nextOffsets)) =>
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: can you rename batchId to something more descriptive so that we can semantically differentiate it from the `currentBatchId`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala [hunk 15]
```diff
@@ -373,22 +380,66 @@ class StreamExecution(
...
+         */
+        currentBatchId = commitLog.getLatest() match {
+          case Some((committedBatchId, _))
+            if batchId == committedBatchId => committedBatchId + 1
+          case _ => batchId
+        }
+        if (batchId < currentBatchId) {
+          /* The last batch was successfully committed, so we can safely process a
+           * new next batch but first:
+           * Make a call a call to getBatch using the offsets from previous batch.
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: "a call" is present twice.

Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: also when referring to getBatch, use source.getBatch to be more clear.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala [hunk 16]
```diff
@@ -373,22 +380,66 @@ class StreamExecution(
...
+        /* identify the current batch id: if commit log indicates we successfully processed the
+         * latest batch id in the offset log, then we can safely move to the next batch
+         * i.e., committedBatchId + 1
+         */
+        currentBatchId = commitLog.getLatest() match {
+          case Some((committedBatchId, _))
+            if batchId == committedBatchId => committedBatchId + 1
+          case _ => batchId
+        }
+        if (batchId < currentBatchId) {
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: The above match-case, and this if statement essentially are the same semantic conditions - both will be true or both will be false. So might as well merge these two into a single if condition.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/Trigger.scala [hunk 17]
```diff
@@ -38,6 +38,51 @@ sealed trait Trigger
...
+ * }}}
+ *
+ * @since 2.2.0
+ */
+@Experimental
+@InterfaceStability.Evolving
+case class OneTime() extends Trigger
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/Trigger.scala]: You need python APIs as well.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryListenerSuite.scala [hunk 18]
```diff
@@ -109,8 +109,9 @@ class StreamingQueryListenerSuite extends StreamTest with BeforeAndAfter {
 
         // Termination event generated with exception message when stopped with error
         StartStream(ProcessingTime(100), triggerClock = clock),
+        AdvanceManualClock(100), // advance clock to ensure completed initial trigger
```
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryListenerSuite.scala]: why is this needed?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingAggregationSuite.scala [hunk 19]
```diff
@@ -327,7 +328,8 @@ class StreamingAggregationSuite extends StateStoreMetricsTest with BeforeAndAfte
         true
       },
       StartStream(ProcessingTime("10 day"), triggerClock = clock),
-      CheckLastBatch((20L, 1), (85L, 1)),
+      // Commit log should prevent batch from running again
```
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingAggregationSuite.scala]: I think we should still check whether re-executing the last batch uses the watermark from the offset log correctly or not.

So i suggest deleting the commit log to force execution of the last batch.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingAggregationSuite.scala [hunk 20]
```diff
@@ -277,7 +277,8 @@ class StreamingAggregationSuite extends StateStoreMetricsTest with BeforeAndAfte
         true
       },
       StartStream(ProcessingTime("10 seconds"), triggerClock = clock),
-      CheckLastBatch((20L, 1), (85L, 1)),
+      // The commit log should ensure that we do not run another batch
+      CheckLastBatch(),
```
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingAggregationSuite.scala]: Same comment as below.

**Comment by tdas** [comment]
also remove the WIP from the title.

File: python/pyspark/sql/streaming.py [hunk 21]
```diff
@@ -315,6 +315,18 @@ def _to_java_trigger(self, sqlContext):
             self.interval)
 
 
+class OneTime():
+    """A trigger that runs a query once and then exits.
+
+    .. note:: Experimental
+
+    .. versionadded:: 2.1
```
Comment (tdas) [inline python/pyspark/sql/streaming.py]: this has to be version 2.2

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala [hunk 22]
```diff
@@ -291,10 +299,13 @@ class StreamExecution(
                   runBatch(sparkSessionToRunBatches)
                 }
               }
-
               // Report trigger as finished and construct progress object.
               finishTrigger(dataAvailable)
               if (dataAvailable) {
+                // Update committed offsets.
+                committedOffsets ++= availableOffsets
+                logDebug(s"Commit log write ${currentBatchId}")
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: i think this debug statement isnt too useful. rather you can generalize it to something like "batch X committed"

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala [hunk 23]
```diff
@@ -392,12 +400,26 @@ class StreamExecution(
    *  - currentBatchId
    *  - committedOffsets
    *  - availableOffsets
+   *  The basic structure of this method is as follows:
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: really like this explanation.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala [hunk 24]
```diff
@@ -419,14 +441,44 @@ class StreamExecution(
             SQLConf.SHUFFLE_PARTITIONS.key, shufflePartitionsToUse.toString)
         }
 
-        logDebug(s"Found possibly unprocessed offsets $availableOffsets " +
-          s"at batch timestamp ${offsetSeqMetadata.batchTimestampMs}")
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: why remove this debug log?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala [hunk 25]
```diff
@@ -419,14 +441,44 @@ class StreamExecution(
...
+        offsetLog.get(latestBatchId - 1).foreach { lastOffsets =>
+          committedOffsets = lastOffsets.toStreamProgress(sources)
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: add a comment on what you are trying to do here. also if this is part of the `// First assume that we are re-executing the latest batch in the offset log` then might as well move this up there if possible.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala [hunk 26]
```diff
@@ -419,14 +441,44 @@ class StreamExecution(
...
+                source.getBatch(start, end)
+              case _ =>
+            }
+            currentBatchId = completionBatchId + 1
+            committedOffsets ++= availableOffsets
+            // Construct a new batch be recomputing availableOffsets
+            constructNextBatch()
+          }
+          case Some((completionBatchId, _)) if completionBatchId + 1 != latestBatchId =>
+            logWarning(s"batch completion log latest batch id is ${completionBatchId}, " +
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: we generally start log messages with caps.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala [hunk 27]
```diff
@@ -419,14 +441,44 @@ class StreamExecution(
...
-        offsetLog.get(batchId - 1).foreach {
-          case lastOffsets =>
-            committedOffsets = lastOffsets.toStreamProgress(sources)
-            logDebug(s"Resuming with committed offsets: $committedOffsets")
+        /* identify the current batch id: if commit log indicates we successfully processed the
+         * latest batch id in the offset log, then we can safely move to the next batch
+         * i.e., committedBatchId + 1
+         */
+        batchCommitLog.getLatest() match {
+          case Some((completionBatchId, _))
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: I would write the conditions as 
```
case Some((completionBatchId, _)) => 
  if (completionBatchId == latestBatchId) { ... }
  else if (completionBatchId < latestBatchId - 1) { logWarning(...) }
case None => 
```
It is less intuitive to look at the current cases and see whether they are exhaustive or not.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala [hunk 28]
```diff
@@ -419,14 +441,44 @@ class StreamExecution(
...
+            if latestBatchId == completionBatchId => {
+            /* The last batch was successfully committed, so we can safely process a
+             * new next batch but first:
+             * Make a call to getBatch using the offsets from previous batch.
+             * because certain sources (e.g., KafkaSource) assume on restart the last
+             * batch will be executed before getOffset is called again.
+             */
+            availableOffsets.foreach {
+              case (source, end)
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: remove case, use `foreach { (source, end) =>  ... }`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala [hunk 29]
```diff
@@ -419,14 +441,44 @@ class StreamExecution(
...
+                if committedOffsets.get(source).map(_ != end).getOrElse(true) =>
+                val start = committedOffsets.get(source)
+                logDebug(s"Initializing offset retrieval from $source " +
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: incorrect. you are not doing offset retrieval. rather say something like getting latest batch from sources but not executing.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala [hunk 30]
```diff
@@ -419,14 +441,44 @@ class StreamExecution(
...
+            committedOffsets ++= availableOffsets
+            // Construct a new batch be recomputing availableOffsets
+            constructNextBatch()
+          }
+          case Some((completionBatchId, _)) if completionBatchId + 1 != latestBatchId =>
+            logWarning(s"batch completion log latest batch id is ${completionBatchId}, " +
+              s"which is not trailing batchid $latestBatchId by one")
+          case _ => logInfo("no commit log present")
         }
+        logDebug(s"Resuming with committed offsets $committedOffsets " +
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: Can you also print the batch id.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/Trigger.scala [hunk 31]
```diff
@@ -38,6 +38,51 @@ sealed trait Trigger
...
+
+  /**
+   * Create a [[OneTime]] trigger.
+   *
+   * Example:
+   * {{{
+   *   df.write.trigger(OneTime.create())
+   * }}}
+   *
+   * @since 2.0.0
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/Trigger.scala]: fix version to 2.2

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/Trigger.scala [hunk 32]
```diff
@@ -38,6 +38,51 @@ sealed trait Trigger
...
+@Experimental
+@InterfaceStability.Evolving
+object OneTime {
+
+  /**
+   * Create a [[OneTime]] trigger.
+   *
+   * Example:
+   * {{{
+   *   df.write.trigger(OneTime.create())
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/Trigger.scala]: df.writeStream
(also fix if this is present anywhere else)

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/Trigger.scala [hunk 33]
```diff
@@ -38,6 +38,51 @@ sealed trait Trigger
...
+ *
+ * @since 2.2.0
+ */
+@Experimental
+@InterfaceStability.Evolving
+case class OneTime() extends Trigger
+
+/**
+ * :: Experimental ::
+ * Used to create [[OneTime]] triggers for [[StreamingQuery]]s.
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/Trigger.scala]: Explain that does one time mean.

**Review by tdas — CHANGES_REQUESTED** [review]
There minor since versions and logging that need to be fixed. But I realized that there are no python tests for python trigger objects. We should add them as part of this PR.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/Trigger.scala (hunk 3 above)
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/Trigger.scala]: yes. this doesnt. please fix them.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala (hunk 24 above)
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: Yeah. this does not make sense here any more. But please add similar logging of recovered metadata later, where you have logged the start and available offsets.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/BatchCommitLog.scala [hunk 34]
```diff
@@ -0,0 +1,80 @@
...
+  override protected def deserialize(in: InputStream): String = {
+    // called inside a try-finally where the underlying stream is closed in the caller
+    val lines = IOSource.fromInputStream(in, UTF_8.name()).getLines()
+    if (!lines.hasNext) {
+      throw new IllegalStateException("Incomplete log file in the offset commit log")
+    }
+    val version = lines.next().trim.toInt
+    if (BatchCommitLog.VERSION < version) {
+      throw new IllegalStateException(s"Incompatible log file version ${version}")
+    }
```
Comment (lw-lin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/BatchCommitLog.scala]: nit: now that https://github.com/apache/spark/pull/17070 has been merged, maybe let's use `parseVersion(lines.next(), BatchCommitLog.VERSION)`

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/Trigger.scala (hunk 33 above)
Comment (tcondie) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/Trigger.scala]: The explanation of OneTime trigger is given in the OneTime class definition. Does that suffice, or should the explanation be reiterated in the object definition?

If it should be reiterated, then I can do the same with ProcessingTime.

File: python/pyspark/sql/streaming.py [hunk 35]
```diff
@@ -786,9 +798,8 @@ def queryName(self, queryName):
         self._jwrite = self._jwrite.queryName(queryName)
         return self
 
-    @keyword_only
```
Comment (tdas) [inline python/pyspark/sql/streaming.py]: Removed keyword_only, otherwise its weird if we have to write 
`writeStream.trigger(trigger=OneTime())`

File: python/pyspark/sql/streaming.py [hunk 36]
```diff
@@ -786,9 +798,8 @@ def queryName(self, queryName):
...
     @since(2.0)
-    def trigger(self, processingTime=None):
+    def trigger(self, trigger=None, processingTime=None):
```
Comment (tdas) [inline python/pyspark/sql/streaming.py]: Added this before processingTime so that we can use positional param and  write `trigger(OneTime())` 
Does not break existing APIs.

**Comment by tdas** [comment]
Jenkins test this please

**Comment by sameeragarwal** [comment]
Seems like this broke the 2.10 builds:
[spark-master-compile-maven-scala-2.10 #3992](https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Compile/job/spark-master-compile-maven-scala-2.10/3992/)
[spark-master-compile-sbt-scala-2.10 #4077](https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Compile/job/spark-master-compile-sbt-scala-2.10/4077/)
